scopes/compilation/compiler/workspace-compiler.ts

```diff
@@ -280,14 +282,52 @@ export class WorkspaceCompiler {
     }
   }
 
+  private async getGraph(components: Component[]): Promise<Graph<Component, string>> {
+    const legacyIdStrToComponentId: { [bitIdStr: string]: ComponentID } = {};
+    const graph = new Graph<Component, string>();
+
+    for (const comp of components) {
+      graph.setNode(new Node(comp.id.toString(), comp));
+    }
+
+    for (const comp of components) {
+      const consumerComp = comp.state._consumer as ConsumerComponent;
+      for (const [depType, depIds] of Object.entries(consumerComp.depsIdsGroupedByType)) {
+        for (const depId of depIds) {
+          const id = depId.toString();
+          if (!legacyIdStrToComponentId[id]) {
+            // eslint-disable-next-line no-await-in-loop
+            legacyIdStrToComponentId[id] = await this.workspace.resolveComponentId(depId);
+          }
+          const depCompId = legacyIdStrToComponentId[depId.toString()];
+          if (graph.hasNode(depCompId.toString())) {
+            graph.setEdge(new Edge(comp.id.toString(), depCompId.toString(), depType));
+          }
+        }
+      }
+    }
+    return graph;
+  }
+
   async compileComponents(
     componentsIds: string[] | BitId[] | ComponentID[], // when empty, it compiles new+modified (unless options.all is set),
     options: CompileOptions,
     noThrow?: boolean
   ): Promise<BuildResult[]> {
     if (!this.workspace) throw new ConsumerNotFound();
     const componentIds = await this.getIdsToCompile(componentsIds, options.changed);
-    const components = await this.workspace.getMany(componentIds);
+    let components = await this.workspace.getMany(componentIds);
+
+    // create a graph and sort it to make sure that we compile in the right order (dependencies first)
+    const graph = await this.getGraph(components);
+    try {
+      const sortedSeeders = graph.toposort(true);
+      components = sortedSeeders.map((node) => node.attr);
+    } catch (err) {
+      if (err instanceof CyclicError) {
+        this.logger.consoleWarning(`Failed to sort components for compilation: ${err.message}`);
+      }
+    }
```

Comment:
do not build and sort the graph here, it should be the env responsibility.
We can add another API to the workspace compiler (something like transpileAllComponents) and bit will make sure to pass the entire list of components to the env, so the env can sort them if he needs to.